iree/turbine/kernel/wave/wave.py

```diff
@@ -294,21 +306,41 @@ def is_reduction(node: fx.Node):
                 if tiling_constraint.dim == custom.axis:
                     tiling_constraint.induction_var = self.induction_vars[custom]
 
-    def initialize_wave_constraints(self, trace: CapturedTrace) -> None:
+    def initialize_implicit_constraints(self, trace: CapturedTrace) -> None:
         """
-        For each wave constraint, determines the appropriate wave id by looking
-        for workgroup constraints along the same dimension and using information
-        from the hardware constraints.
+        Initialize all implicit constraints. Add wave constraints to the
+        [x, y, z] dimensions using information from the hardware constraints.
 
         """
 
         hardware_constraint = self.hardware_constraints[0]
-        for wave_constraint in self.wave_constraints:
-            for workgroup_constraint in self.workgroup_constraints:
-                if wave_constraint.dim == workgroup_constraint.dim:
-                    wave_constraint.set_wave_id_from_hardware_and_workgroup_constraint(
-                        hardware_constraint, workgroup_constraint
-                    )
+        constraints = []
+        for workgroup_constraint in self.workgroup_constraints:
+            # Ignore `workgroup_dim > 2` and dimensions with only one thread (ie. the value is always 0).
+            if (
+                workgroup_constraint.workgroup_dim > 2
```

Comment:
Styling NIT: I'd recommend for this massive if condition to be broken apart for better legibility

Reply:
Once, we break apart the if(s), I'd also add a comment wrt `workgroup_dim !=0 and waves_per_block == 1` with something simlar to https://github.com/iree-org/iree-turbine/blob/de8f92ebcf86058338efe5833d4d863a3a636e44/iree/turbine/kernel/wave/constraints.py#L493-L495